src/net/p2p.rs

```diff
@@ -3028,11 +3050,355 @@ impl PeerNetwork {
+                        network.mempool_sync_reset();
+                        return Ok(true);
+                    }
+                }
+                Some(ref mut convo) => {
+                    match convo.try_get_response() {
+                        None => {
+                            // still waiting
+                            debug!(
+                                "Mempool sync event {} still waiting for a response",
+                                event_id
+                            );
+                            return Ok(false);
+                        }
+                        Some(http_response) => match http_response {
+                            HttpResponseType::MemPoolTxs(_, txs) => {
+                                network.mempool_sync_transactions = Some(txs);
+                                return Ok(true);
+                            }
+                            _ => {
+                                warn!("Mempool sync request received {:?}", &http_response);
+                                network.mempool_sync_reset();
+                                return Ok(true);
+                            }
+                        },
+                    }
+                }
+            }
+        })
+    }
+
+    /// Do a mempool sync
+    /// Return true if we're done and can advance to the next state.
+    fn do_mempool_sync(
+        &mut self,
+        dns_client_opt: &mut Option<&mut DNSClient>,
+        mempool: &MemPoolDB,
+        chainstate: &mut StacksChainState,
+    ) -> Result<bool, net_error> {
+        if get_epoch_time_secs() <= self.mempool_sync_deadline {
+            debug!(
+                "Wait until {} to do a mempool sync",
+                self.mempool_sync_deadline
+            );
+            return Ok(true);
+        }
+
+        if self.mempool_sync_timeout == 0 {
+            // begin new sync
+            self.mempool_sync_timeout =
+                get_epoch_time_secs() + self.connection_opts.mempool_sync_timeout;
+        } else {
+            if get_epoch_time_secs() > self.mempool_sync_timeout {
+                debug!("Mempool sync took too long; terminating");
+                self.mempool_sync_reset();
+                return Ok(true);
+            }
+        }
+
+        // 1. pick a random outbound conversation.
+        // Sets self.mempool_sync_data_url on success.
+        if self.mempool_sync_data_url.is_none() {
+            if self.mempool_sync_pick_outbound_peer(dns_client_opt)? {
+                return Ok(true);
+            }
+        }
+
+        // 2. resolve its data URL
+        // sets self.mempool_sync_sockaddr on success
+        if self.mempool_sync_sockaddr.is_none() {
+            if let Some(url_str) = self.mempool_sync_data_url.clone() {
+                if self.mempool_sync_resolve_data_url(&url_str, dns_client_opt)? {
+                    return Ok(true);
+                }
+            }
+        }
```

Comment:
The control flow of these if statements seems like it could lead to some unintended paths. For example:

```
self.mempool_sync_data_url = None
self.mempool_sync_pick_outbound_peer() => returns false
```

The control flow would then check all the other if statements in this method. Presumably they all return false (because at some point `self.mempool_sync_reset` was invoked), and so this method returns `Ok(false)`, but it seems like an unsafe pattern. In most of these cases, does it make sense to just return the inner function response, e.g.,

```rust
        if self.mempool_sync_data_url.is_none() {
            return self.mempool_sync_pick_outbound_peer(dns_client_opt);
        }
```

?

Reply:
So, I admit I was a bit lazy here.  The more robust way of doing this would be to have an explicit `mempool_state` that indicated which state the mempool synchronizer was actually in.  Right now, the code paths here do the same work as having an explicit `mempool_state`, but implicitly.  Instead, I can have this code just run off a `match self.mempool_state {}` block that updates `self.mempool_state`, so that it's always clear what the state machine should be doing in reaction to attempted state transitions (like how we do in `do_network_work()`).

Reply:
I went ahead and refactored this so that there's now an explicit state machine loop for the mempool, and each state-transition function will return the data it generated instead of trying to store it to the `PeerNetwork`.  This way, all state-transition code and all state book-keeping (including resets) happen exclusively in the top-level mempool sync state machine loop.